File: 백준/Silver/15654. N과 M （5）/N과 M （5）.js
const fs = require("fs");
const path = process.platform === "linux" ? "/dev/stdin" : "./input.txt";

const input = fs
  .readFileSync(path, "utf8")
  .split("\n")
  .map((item) => item.trim());

const [N, M] = input[0].split(" ").map((item) => parseInt(item));

const numbers = input[1].split(" ").map((item) => parseInt(item));
numbers.sort((a, b) => a - b);

const visited = Array(M + 1).fill(false);
const answer = [];

function backtrack(current) {
  if (current.length === M) {
    answer.push(current.join(" "));
    return;
  }

  for (let i = 0; i < N; i++) {
    if (!visited[i]) {
      visited[i] = true;
      current.push(numbers[i]);
      backtrack(current);
      current.pop();
      visited[i] = false;
    }
  }
}

backtrack([]);
console.log(answer.join("\n"));

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.py
import sys

input = sys.stdin.readline

N, M = map(int, input().split(' '))

arr = list(map(int, input().split(' ')))
arr.sort()
visited = [False for _ in range(N)]

result = []

def dfs():
    if len(result) == M:
        print(*result)
        return

    for i in range(0, N):
        if not visited[i]:
            visited[i] = True
            result.append(arr[i])
            dfs()
            visited[i] = False
            result.pop()

dfs()